### PR TITLE
Make ArrayPoolList.GuardIndex inlinable

### PR DESCRIPTION
**Pull request by benaadams**
Heavily called, however doesn't inline due to the large amount of code for throwing the exceptions

Before

image

After
image

## Changes

- Move exceptions out to non-returning throw helpers, so aren't included in the inline
- Use cast to `uint` to do less comparisons
- Delete `ByteArrayPoolList` as unused and seal the class

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [ ] Yes
- [x] No

#### If yes, did you write tests?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Core/Collections/ArrayPoolList.cs [hunk 1]
```diff
@@ -160,20 +170,23 @@ private bool RemoveAtInternal(int index, bool shouldThrow)
...
             GuardDispose();
-            if (index < 0)
-            {
-                return shouldThrow
-                    ? throw new ArgumentOutOfRangeException($"Index {index} is below 0.")
-                    : false;
-            }
-            else if (index >= _count && (!allowEqualToCount || index > _count))
+            int count = _count;
+            if ((uint)index > (uint)count || (!allowEqualToCount && index == count))
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Collections/ArrayPoolList.cs:174]: can we ignore `index < 0` here?

Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Collections/ArrayPoolList.cs:174]: Cast to `uint` means any negative `int`s will become higher than `int.MaxValue` so it gets covered in the greater than check
`(uint)index > (uint)count` as `count` can't be that high

Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Collections/ArrayPoolList.cs:174]: Is a common pattern in the runtime to only have a single in range check; without testing for negative values [Span.cs#L148-L159](https://github.com/dotnet/runtime/blob/be84df7642f21702b0e7859f0c134e4d6cfc67eb/src/libraries/System.Private.CoreLib/src/System/Span.cs#L148-L159)

image

**Review by Scooletz — APPROVED** [review]
👋 calls! Nice 💪

**Pull request merged**